### PR TITLE
Support custom credential provider

### PR DESCRIPTION
**Pull request by JusSergey**
Added new options:
1. Removed auto-refresh for role-based credentials
    I discovered that this is already implemented in
    `AssumeRoleProvider`.
2. Implement the `CustomCredentialsProvider` trait
    This trait adds the ability to implement providing the credentials
    outside of the `ssstar`, the user of this crate can implement this trait
    and pass it into to use a custom way to get the credentials

This is a part of the fix of the issue https://github.com/elastio/elastio/issues/6796

### CODE REVIEW

**Comment by anelson** [comment]
There's no link to an issue, or any description of what this is or why it's useful.  Please elaborate on what this is for.

File: ssstar/src/role_credentials_provider.rs [hunk 1]
```diff
@@ -14,8 +14,22 @@ use std::{
...
+/// The configuration to assume bastion-based role
+#[derive(Debug)]
+struct BastionConfig {
+    /// Role-arn of bastion role
+    role_arn: String,
+    /// The name of the SSM external-ID parameter
+    external_id_name: String,
+    /// Session name for assuming the bastion-based role
+    session_name: String,
+}
```
Comment (Veetaha) [inline ssstar/src/role_credentials_provider.rs:26]: The concept of `bastion` role is private to `elastio` repository. It shouldn't be exposed in the `ssstar` crate. `ssstar` shouldn't even mention the name "bastion". Instead, it should accept the following parameters for the role to assume:
- ARN of the role
- Value of the external ID parameter (yes, the caller needs to resolve the external ID on its own); we need to modify our `asset-account-auth` crate in `elastio/elastio` repository to expose that information in the `ConnectorIdentity`
- Name of the session

Because `ssstar` is a generic tool this should be named just "role to assume", and not "bastion role to assume with eternal ID from this SSM parameter"; Convenience wrappers for working with the bastion role may be implemented in the `elastio/account`. Btw. the logic of assuming the `bastion` role may change in the future and it's better to reuse it. I think it's better to discuss this problem on a call. Please ping me when you have time for this

**Review by Veetaha — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -11,7 +11,7 @@ keywords      = ["tar", "s3", "archive"]
 license       = "Apache-2.0 OR MIT"
 readme        = "README.md"
 repository    = "https://github.com/elastio/ssstar"
-version       = "0.5.0"
+version       = "0.5.1"
```
Comment (anelson) [inline Cargo.toml:14]: Updating the crate versions is part of the release process, not part of the PR.  Revert the changes to the crate versions.

File: ssstar/src/role_credentials_provider.rs (hunk 1 above)
Comment (JusSergey) [reply ssstar/src/role_credentials_provider.rs]: I entirely removed this from ssstar, instead, i've added a new trait `CustomCredentialsUpdateCallback` which the user should implement and pass to ssstar

Comment (JusSergey) [reply ssstar/src/role_credentials_provider.rs]: @Veetaha I thought the `bastion` is something common to AWS because i've found a post https://docs.aws.amazon.com/managedservices/latest/userguide/using-bastions.html

but probably in our case it's something specific

File: Cargo.toml (hunk 2 above)
Comment (JusSergey) [reply Cargo.toml]: reverted. But I thought it already has to have version `0.5.1` because the `0.5.1` is already released but in master `ssstar` crate has version `0.5.0`

File: ssstar/src/custom_credentials_provider.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            CustomCredentialsUpdateOutput,
+            Box<dyn std::error::Error + Send + Sync + 'static>,
+        >,
+    >;
+}
+
+/// Custom credentials provider that actually implements `ProvideCredentials` and implements the needed
+/// traits to make it possible to put it into config as an option
+#[derive(Debug, Clone)]
+pub struct CustomCredentialsProvider(Arc<Box<dyn CustomCredentialsUpdateCallback>>);
```
Comment (Veetaha) [inline ssstar/src/custom_credentials_provider.rs:38]:
```suggestion
pub struct CustomCredentialsProvider(Arc<dyn CustomCredentialsUpdateCallback>);
```
`Arc<Box>` never makes sense. Just use only Arc

File: ssstar/src/custom_credentials_provider.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+/// This represents the credentials to access to AWS API.
+#[derive(Clone, Debug)]
+pub struct CustomCredentialsUpdateOutput {
+    pub access_key: String,
+    pub secret_key: String,
+    pub session_token: Option<String>,
+}
+
+/// This is a trait that represents a callback to update credentials when they become expired
+pub trait CustomCredentialsUpdateCallback: Debug + Send + Sync {
```
Comment (Veetaha) [inline ssstar/src/custom_credentials_provider.rs:21]: Why not to use `#[async_trait]`?

File: ssstar/src/custom_credentials_provider.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+/// because this must be never called
+impl PartialEq for CustomCredentialsProvider {
+    fn eq(&self, _other: &Self) -> bool {
+        unreachable!("BUG: You should never compare `CustomCredentialsProvider`");
+    }
+}
+
+impl CustomCredentialsProvider {
+    pub fn new(inner: impl CustomCredentialsUpdateCallback + 'static) -> Self {
+        Self(Arc::new(Box::new(inner)))
```
Comment (Veetaha) [inline ssstar/src/custom_credentials_provider.rs:60]:
```suggestion
        Self(Arc::new(inner))
```

File: ssstar/src/config.rs [hunk 6]
```diff
@@ -74,7 +76,11 @@ pub struct Config {
         feature = "clap",
         clap(long, global = true, conflicts_with_all = &["aws_access_key_id", "aws_secret_access_key", "aws_session_token"])
     )]
-    pub aws_role_session_duration_seconds: Option<i32>,
+    pub aws_role_session_duration_seconds: Option<u64>,
+
+    /// Custom credentials provider, this is useful if you prefer providing credentials on yourself
+    /// which may expire from time to time.
+    pub credentials_provider: Option<CustomCredentialsProvider>,
```
Comment (Veetaha) [inline ssstar/src/config.rs:83]: Is it possible to put smth like `#[clap(skip)]` here to remove the requirements for implementing `FromStr`, and `PartialEq`?

```suggestion
    #[clap(skip)]
    pub credentials_provider: Option<CustomCredentialsProvider>,
```

Comment (JusSergey) [reply ssstar/src/config.rs:83]: it helps, now I don't need to implement `FromStr` but it still requires `PartialEq`, I think `clap` can't resolve the issue that it requires `PartialEq` because it doesn't related to clap, it's needed because we have a test where we compare the default variant of config with expected variant

File: ssstar/src/custom_credentials_provider.rs (hunk 4 above)
Comment (JusSergey) [reply ssstar/src/custom_credentials_provider.rs:21]: i just automatically written it without `async_trait`, it didn't look so bad to me. Re-written with `async_trait`.

File: ssstar/src/custom_credentials_provider.rs (hunk 5 above)
Comment (JusSergey) [reply ssstar/src/custom_credentials_provider.rs:60]: fixed

File: ssstar/src/custom_credentials_provider.rs (hunk 3 above)
Comment (JusSergey) [reply ssstar/src/custom_credentials_provider.rs:38]: you're right, probably I added `Box` when there' wasn `Arc`. Fixed

File: Cargo.toml (hunk 2 above)
Comment (anelson) [reply Cargo.toml]: If so that's an issue w/ the release process.  It shouldn't be patched around in this PR.  Let's revisit that when it's time to release this change.

**Review by anelson — APPROVED** [review]

File: ssstar/src/config.rs (hunk 6 above)
Comment (Veetaha) [reply ssstar/src/config.rs:87]: I see, makes sense, thank you for explanation

**Review by Veetaha — APPROVED** [review]

**Comment by JusSergey** [comment]
@anelson I fixed CI checks error, it failed because of a new version of `cargo +nightly`, to fix it I hardcoded specific version of nightly to `+nightly-2023-06-15`, with this version the CI checks pass, with the old version of nightly, udeps failed to like that:
Screenshot from 2023-07-17 13-34-58

I found that somebody commented in an github's issue that it reproduces on the latest version of nightly https://github.com/dtolnay/proc-macro2/issues/356#issuecomment-1614449061

and someone else suggests downgrading the version of nightly https://github.com/dtolnay/proc-macro2/issues/356#issuecomment-1614467505

I just want to make sure that it is okay

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -27,7 +27,7 @@ jobs:
     - name: Install Rust toolchain
       uses: actions-rs/toolchain@v1
       with:
-        toolchain: nightly
+        toolchain: nightly-2023-06-15
```
Comment (anelson) [inline .github/workflows/ci.yml:30]: I don't love this change, but I don't have a better proposal.  Please put a comment block here, with the same explanation you wrote in the PR comment, so we don't lose track of why this very specific nightly version is pinned here.
